Reqnroll/TestRunnerManager.cs

```diff
@@ -173,15 +209,32 @@ public virtual async Task DisposeAsync()
         {
             await FireTestRunEndAsync();
 
-            foreach (var testRunner in _testRunnerRegistry.Values)
+            if (_globalTestRunner != null)
             {
-                testRunner.TestThreadContext.TestThreadContainer.Dispose();
+                ReleaseTestThreadContext(_globalTestRunner.TestThreadContext);
             }
 
-            // this call dispose on this object, but the disposeLockObj will avoid double execution
+            var items = _availableTestWorkerContainers.ToArray();
+            while (items.Length > 0)
+            {
+                foreach (var item in items)
+                {
+                    item.Key.Dispose();
+                    _availableTestWorkerContainers.TryRemove(item.Key, out _);
+                }
+                items = _availableTestWorkerContainers.ToArray();
+            }
+
+            var notReleasedRunner = _usedTestWorkerContainers.ToArray();
+            if (notReleasedRunner.Length > 0)
+            {
+                var errorText = $"Found {notReleasedRunner.Length} not released TestRunners (ids: {string.Join(",", notReleasedRunner.Select(x => TestThreadContainerInfo.GetId(x.Key)))})";
+                _globalContainer.Resolve<ITestTracer>().TraceWarning(errorText);
```

Comment:
@gasparnagy I'm unsure if this is the correct way to report a warning this late in the tests. Are there better alternatives?

Reply:
Good question. We do similar thing at `ContextManager.Cleanup` that is also "late", so I would say yes, that is the best we can do. But it might not be shown for some test runners.

Reply:
Okay if this is the best known way, it's good. 🙂